File: backend/app/services/auth_service.py
"""账号服务：手机号自动注册/登录、密码注册、登录态维护。"""
import logging
from datetime import datetime, timezone

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.security import hash_password, verify_password
from app.models.user import User

logger = logging.getLogger("app.auth")

# 内存验证码：{phone: code}（仅演示；多实例/生产应使用 Redis 并设 TTL）
_sms_store: dict[str, str] = {}


def issue_sms_code(phone: str) -> str:
    from app.core.config import settings

    code = settings.MOCK_SMS_CODE
    _sms_store[phone] = code
    # mock：不接第三方短信，仅输出日志便于演示
    logger.info("[mock-sms] 向 %s 发送验证码：%s", phone, code)
    return code


def check_sms_code(phone: str, code: str) -> bool:
    expected = _sms_store.get(phone)
    return bool(expected) and expected == code


def get_or_create_user(db: Session, phone: str, *, role: str = "learner", nickname: str | None = None) -> User:
    user = db.scalar(select(User).where(User.phone == phone))
    if user is None:
        user = User(
            phone=phone,
            user_role=role,
            age_group="adult",
            nickname=nickname or f"学习者{phone[-4:]}",
            status=1,
        )
        db.add(user)
        db.flush()
        logger.info("注册新账号 phone=%s role=%s id=%s", phone, role, user.user_id)
    return user


def authenticate_by_password(db: Session, phone: str, password: str) -> User | None:
    user = db.scalar(select(User).where(User.phone == phone))
    if user is None or not user.password_hash:
        return None
    return user if verify_password(password, user.password_hash) else None


def set_password(user: User, password: str) -> None:
    user.password_hash = hash_password(password)


def touch_login(user: User) -> None:
    user.last_login_time = datetime.now(timezone.utc).replace(tzinfo=None)

File: backend-spring/src/main/java/com/englishlearn/service/AuthService.java
package com.englishlearn.service;

import com.englishlearn.entity.User;
import com.englishlearn.repository.UserRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * 账号服务：手机号自动注册/登录、密码注册、登录态维护（对齐 FastAPI auth_service）。
 */
@Service
public class AuthService {

    private static final Logger log = LoggerFactory.getLogger(AuthService.class);
    private static final Map<String, String> SMS_STORE = new ConcurrentHashMap<>();

    private final UserRepository userRepository;
    private final BCryptPasswordEncoder passwordEncoder;
    private final String mockSmsCode;

    public AuthService(UserRepository userRepository,
                       BCryptPasswordEncoder passwordEncoder,
                       @Value("${app.sms.mock-code:123456}") String mockSmsCode) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.mockSmsCode = mockSmsCode;
    }

    public boolean phoneValid(String phone) {
        return phone != null && phone.chars().allMatch(Character::isDigit)
                && phone.length() >= 6 && phone.length() <= 20;
    }

    public String issueSmsCode(String phone) {
        SMS_STORE.put(phone, mockSmsCode);
        log.info("[mock-sms] 向 {} 发送验证码：{}", phone, mockSmsCode);
        return mockSmsCode;
    }

    public boolean checkSmsCode(String phone, String code) {
        String expected = SMS_STORE.get(phone);
        return expected != null && expected.equals(code);
    }

    @Transactional
    public User getOrCreateUser(String phone, String role, String nickname) {
        User user = userRepository.findByPhone(phone).orElse(null);
        if (user == null) {
            user = new User();
            user.phone = phone;
            user.userRole = role;
            user.ageGroup = "adult";
            user.nickname = (nickname == null || nickname.isBlank())
                    ? "学习者" + phone.substring(phone.length() - 4)
                    : nickname;
            user.status = 1;
            user.registerTime = LocalDateTime.now();
            user = userRepository.save(user);
            log.info("注册新账号 phone={} role={} id={}", phone, role, user.userId);
        }
        return user;
    }

    public User authenticateByPassword(String phone, String password) {
        User user = userRepository.findByPhone(phone).orElse(null);
        if (user == null || user.passwordHash == null) {
            return null;
        }
        return passwordEncoder.matches(password, user.passwordHash) ? user : null;
    }

    public boolean verifyPassword(String plain, String hashed) {
        return hashed != null && passwordEncoder.matches(plain, hashed);
    }

    public String hashPassword(String plain) {
        return passwordEncoder.encode(plain);
    }

    public void setPassword(User user, String password) {
        user.passwordHash = passwordEncoder.encode(password);
    }

    public void touchLogin(User user) {
        user.lastLoginTime = LocalDateTime.now();
    }
}